### PR TITLE
Add test case for snapshot and restore of encrypted RBD PVCs

### PR DESCRIPTION
**Pull request by rachael-george**
Add test case for creating snapshots of encrypted RBD PVCs and restoring them
Signed-off-by: rachael-george <rgeorge@redhat.com>

### CODE REVIEW

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+class TestEncryptedRbdBlockPvcSnapshot(ManageTest):
+    """
+    Tests to take snapshots of encrypted RBD Block VolumeMode PVCs
+
+    """
+
+    @pytest.fixture(autouse=True)
+    def setup(self, kv_version, request):
+        """
+        Setup csi-kms-connection-details configmap
```
[the remark this answers is not in the document]
Comment (ebenahar) [reply tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py:50]: +1. This seems similar to tests.manage.pv_services.test_rbd_pv_encryption.TestRbdPvEncryption.setup

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+                kms.remove_kmsid(self.new_kmsid)
+
+            # Delete the resources in vault
+            self.vault.remove_vault_backend_path()
+            self.vault.remove_vault_policy()
+            self.vault.remove_vault_namespace()
+
+        request.addfinalizer(finalizer)
+
+    def test_encrypted_rbd_block_pvc_snapshot(
```
Comment (ebenahar) [inline tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py:111]: The test case's implementation has some similarities to tests.manage.pv_services.test_rbd_pv_encryption.TestRbdPvEncryption.test_rbd_pv_encryption

Please try to merge the similarities into functions, which will be called from both tests

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rachael-george) [reply tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py:46]: Done

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py (hunk 2 above)
Comment (rachael-george) [reply tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py:53]: Done

File: tests/conftest.py [hunk 3]
```diff
@@ -3868,3 +3868,82 @@ def teardown():
 
     request.addfinalizer(teardown)
     return ripsaw
+
+
+@pytest.fixture(scope="function")
+def pv_encryption_kms_setup_factory(request):
+    """
+    Create vault resources and setup csi-kms-connection-details configMap
+    """
```
Comment (jilju) [inline tests/conftest.py:3877]:
```suggestion
    Create vault resources and setup csi-kms-connection-details configMap

    """
```

File: tests/conftest.py [hunk 4]
```diff
@@ -3868,3 +3868,82 @@ def teardown():
...
+    vault = KMS.Vault()
+
+    def factory(kv_version):
+        """
+        Args:
+            kv_version(str): KV version to be used, either v1 or v2
+        Returns:
+            object: Vault(KMS) object
+        """
```
Comment (jilju) [inline tests/conftest.py:3886]:
```suggestion
            kv_version(str): KV version to be used, either v1 or v2

        Returns:
            object: Vault(KMS) object

        """
```

File: tests/conftest.py [hunk 5]
```diff
@@ -3868,3 +3868,82 @@ def teardown():
...
+            else:
+                new_kmsid = "1-vault"
+                vault.create_vault_csi_kms_connection_details()
+
+        return vault
+
+    def finalizer():
+        """
+        Remove the vault config from csi-kms-connection-details configMap
+        """
```
Comment (jilju) [inline tests/conftest.py:3940]:
```suggestion
        Remove the vault config from csi-kms-connection-details configMap

        """
```

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+@skipif_ocp_version("<4.8")
+class TestEncryptedRbdBlockPvcSnapshot(ManageTest):
+    """
+    Tests to take snapshots of encrypted RBD Block VolumeMode PVCs
+
+    """
+
+    @pytest.fixture(autouse=True)
+    def setup(self, kv_version, pv_encryption_kms_setup_factory):
+        """ "
```
Comment (jilju) [inline tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py:45]:
```suggestion
        """
```

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+        Setup csi-kms-connection-details configmap
+        """
```
Comment (jilju) [inline tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py:47]:
```suggestion
        Setup csi-kms-connection-details configmap

        """
```

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+            if kms.is_key_present_in_path(
+                key=vol_handle, path=self.vault.vault_backend_path
+            ):
+                log.info(f"Vault: Found key for {pvc_obj.name}")
+            else:
+                raise ResourceNotFoundError(f"Vault: Key not found for {pvc_obj.name}")
+
+        log.info(
+            "Check for encrypted device, find initial md5sum value and run IO on all pods"
+        )
```
Comment (jilju) [inline tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py:121]: I think the above steps need not be part of the test case. The test case is verifying snapshot and restore of encrypted PVC. So creation of project , storageclass, pvc and pods can be added in the test setup.

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+        for vol_handle, pod_obj in zip(vol_handles, pod_objs):
+            # Verify whether encrypted device is present inside the pod
+            if pod_obj.exec_sh_cmd_on_pod(
+                command=f"lsblk | grep {vol_handle} | grep crypt"
+            ):
+                log.info(f"Encrypted device found in {pod_obj.name}")
+            else:
+                log.error(f"Encrypted device not found in {pod_obj.name}")
```
Comment (jilju) [inline tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py:129]: Isn't this a cause of test failure ?

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+                )
+
+        # Verify encrypted device is present and md5sum on all pods
+        for vol_handle, pod_obj in zip(restore_vol_handles, restore_pod_objs):
+            if pod_obj.exec_sh_cmd_on_pod(
+                command=f"lsblk | grep {vol_handle} | grep crypt"
+            ):
+                log.info(f"Encrypted device found in {pod_obj.name}")
+            else:
+                log.error(f"Encrypted device not found in {pod_obj.name}")
```
Comment (jilju) [inline tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py:284]: Isn't this a cause of test failure ?

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py (hunk 8 above)
Comment (rachael-george) [reply tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py:127]: Done

File: tests/conftest.py (hunk 3 above)
Comment (rachael-george) [reply tests/conftest.py]: Done

File: tests/conftest.py (hunk 5 above)
Comment (rachael-george) [reply tests/conftest.py]: Done

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py (hunk 6 above)
Comment (rachael-george) [reply tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py]: Done

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py (hunk 7 above)
Comment (rachael-george) [reply tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py]: Done

File: tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py (hunk 9 above)
Comment (rachael-george) [reply tests/manage/pv_services/pvc_snapshot/test_encrypted_rbd_pvc_snapshot.py]: Yes, updated the code

**Review by jilju — APPROVED** [review]

**Review by sidhant-agrawal — APPROVED** [review]

**Review by Akarsha-rai — APPROVED** [review]
LGTM

**Review by shylesh — APPROVED** [review]
+1

**Pull request merged**